// @framer/guischneider.assembly-v2/code/Inputs/Qz_Input.tsx
import * as React from "react"
import { Frame, useCycle, addPropertyControls, ControlType } from "framer"
import { qzYellow, qzCoral, qzInk, qzTeal, qzWhite } from "./Assembly"

// Open Preview (CMD + P)
// API Reference: https://www.framer.com/api

export function Qz_Input(props) {
    return (
        <div>
            <label style={UIInput}>
                <div style={UIInputContent}>
                    <input
                        style={UIInputInput}
                        type={props.inputType}
                        // value={props.value}
                    />
                    <span style={UIInputBorder} />
                </div>
                <span style={UIInputLabel}>{props.label}</span>
            </label>
        </div>
    )
}
Qz_Input.defaultProps = {
    label: "label",
    inputType: "text",
    value: " ",
}
// Add Property Controls
addPropertyControls(Qz_Input, {
    value: {
        title: "Value",
        type: ControlType.String,
        defaultValue: "Enter value",
    },
    label: {
        title: "Label",
        type: ControlType.String,
        defaultValue: "Label",
    },
    inputType: {
        title: "Type",
        type: ControlType.Enum,
        defaultValue: "text",
        options: ["text", "number", "password"],
        optionTitles: ["text", "number", "password"],
    },
})

const UIInput: React.CSSProperties = {
    cursor: "pointer",
    display: "block",
    fontWeight: 400,
    fontSize: "1.125rem",
    lineHeight: 1.2,
    fontFamily: "Avenir",
}
const UIInputContent: React.CSSProperties = {
    WebkitAlignItems: "flex-start",
    alignItems: "flex-start",
    display: "flex",
    WebkitFlexFlow: "row wrap",
    flexFlow: "row wrap",
    position: "relative",
    width: "100%",
}
const UIInputInput: React.CSSProperties = {
    WebkitAppearance: "none",
    MozAppearance: "none",
    appearance: "none",
    backgroundColor: "transparent",
    border: "none",
    borderRadius: "0",
    boxShadow: "none",
    color: "#2d3639",
    cursor: "text",
    WebkitFlex: "1 1 auto",
    flex: "1 1 auto",
    font: "inherit",
    letterSpacing: "inherit",
    lineHeight: "inherit",
    outline: "none",
    paddingBottom: "0.375rem",
    textAlign: "inherit",
    whiteSpace: "pre-wrap",
    width: "100%",
    wordBreak: "break-word",
}
const UIInputBorder: React.CSSProperties = {
    WebkitAlignSelf: "flex-end",
    alignSelf: "flex-end",
    display: "block",
    height: "0.25rem",
    boxShadow: "0 0.125rem 0 0 #455358",
    transition: "all 120ms cubic-bezier(0.47,0,0.745,0.715)",
    width: "100%",
}
const UIInputLabel: React.CSSProperties = {
    color: "#97a5aa",
    display: "block",
    marginTop: "0.625rem",
    textAlign: "inherit",
    fontWeight: 600,
    textTransform: "uppercase",
    fontSize: "0.75rem",
    letterSpacing: "0.0625rem",
    lineHeight: "1.33",
    WebkitUserSelect: "none",
    MozUserSelect: "none",
    userSelect: "none",
}
// function getClassFromStyle(style) {
//     if (style === "normalStyle") return normalStyle
//     if (style === "hoverStyle") return hoverStyle
//     if (style === "disabledStyle") return disabledStyle
//     if (style === "outlinedStyle") return outlinedStyle
//     if (style === "alertStyle") return alertStyle
// }
